In [1]:
#Import libraries

import pandas as pd
import numpy as np
import os

In [2]:
# Set path

path = r'C:\Users\Josh Wattay\OneDrive\Documents\Desktop\Flight Delay Analysis - Raw Data\2022'

In [3]:
# Import Datasets

df = pd.read_csv(os.path.join(path, 'flights2022.csv'), index_col = False)

C:\Users\Josh Wattay\AppData\Local\Temp\ipykernel_14240\2238960500.py:3: DtypeWarning: Columns (11,52) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(os.path.join(path, 'flights2022.csv'), index_col = False)


In [4]:
# Shape Check
df.shape

(7013508, 70)

In [5]:
# Set the display option to show all 70 columns
pd.set_option('display.max_columns', 70)

In [6]:
#Inspect Dataframe and begin Cleaning by ensuring columns are named appropriately
df.head()

Unnamed: 0  Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0           0  2022        1      1           6          4  2022-01-06   
1           1  2022        1      1           6          4  2022-01-06   
2           2  2022        1      1           6          4  2022-01-06   
3           3  2022        1      1           6          4  2022-01-06   
4           4  2022        1      1           6          4  2022-01-06   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        DL                     19790   
1                        DL                     19790   
2                        DL                     19790   
3                        DL                     19790   
4                        DL                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             1581   
1                          DL                             1582   
2                          DL                             1582   
3                          DL                             1583   
4                          DL                             1584   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 DL   
1                                               NaN                 DL   
2                                               NaN                 DL   
3                                               NaN                 DL   
4                                               NaN                 DL   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     19790                          DL      N315DN   
1                     19790                          DL      N545US   
2                     19790                          DL      N545US   
3                     19790                          DL      N345NB   
4                     19790                          DL      N978AT   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                             1581            11697             1169706   
1                             1582            10397             1039707   
2                             1582            11697             1169706   
3                             1583            11697             1169706   
4                             1584            10397             1039707   

   OriginCityMarketID Origin       OriginCityName OriginState  \
0               32467    FLL  Fort Lauderdale, FL          FL   
1               30397    ATL          Atlanta, GA          GA   
2               32467    FLL  Fort Lauderdale, FL          FL   
3               32467    FLL  Fort Lauderdale, FL          FL   
4               30397    ATL          Atlanta, GA          GA   

   OriginStateFips OriginStateName  DestAirportID  DestAirportSeqID  \
0               12         Florida          12953           1295304   
1               13         Georgia          11697           1169706   
2               12         Florida          10397           1039707   
3               12         Florida          14492           1449202   
4               13         Georgia          12448           1244807   

   DestCityMarketID Dest           DestCityName DestState  DestStateFips  \
0             31703  LGA           New York, NY        NY             36   
1             32467  FLL    Fort Lauderdale, FL        FL             12   
2             30397  ATL            Atlanta, GA        GA             13   
3             34492  RDU     Raleigh/Durham, NC        NC             37   
4             32448  JAN  Jackson/Vicksburg, MS        MS             28   

    DestStateName  CRSDepTime  DepTime  DepDelay  DepDelayMinutes  DepDel15  \
0        New York        1126      NaN       NaN              NaN       NaN   
1         Florida        1631   1627.0      -4.0              

The 'Unnamed: 0' column will be removed as it is an additional index and redundant. 

In [7]:
# Drop the 'Unnamed: 0' column
df = df.drop(columns=['Unnamed: 0'])

In [8]:
df.shape

(7013508, 69)

In [9]:
df.head()

Year  Quarter  Month  DayofMonth  DayOfWeek  FlightDate  \
0  2022        1      1           6          4  2022-01-06   
1  2022        1      1           6          4  2022-01-06   
2  2022        1      1           6          4  2022-01-06   
3  2022        1      1           6          4  2022-01-06   
4  2022        1      1           6          4  2022-01-06   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        DL                     19790   
1                        DL                     19790   
2                        DL                     19790   
3                        DL                     19790   
4                        DL                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             1581   
1                          DL                             1582   
2                          DL                             1582   
3                          DL                             1583   
4                          DL                             1584   

  IATA_Code_Originally_Scheduled_Code_Share_Airline Operating_Airline   \
0                                               NaN                 DL   
1                                               NaN                 DL   
2                                               NaN                 DL   
3                                               NaN                 DL   
4                                               NaN                 DL   

   DOT_ID_Operating_Airline IATA_Code_Operating_Airline Tail_Number  \
0                     19790                          DL      N315DN   
1                     19790                          DL      N545US   
2                     19790                          DL      N545US   
3                     19790                          DL      N345NB   
4                     19790                          DL      N978AT   

   Flight_Number_Operating_Airline  OriginAirportID  OriginAirportSeqID  \
0                             1581            11697             1169706   
1                             1582            10397             1039707   
2                             1582            11697             1169706   
3                             1583            11697             1169706   
4                             1584            10397             1039707   

   OriginCityMarketID Origin       OriginCityName OriginState  \
0               32467    FLL  Fort Lauderdale, FL          FL   
1               30397    ATL          Atlanta, GA          GA   
2               32467    FLL  Fort Lauderdale, FL          FL   
3               32467    FLL  Fort Lauderdale, FL          FL   
4               30397    ATL          Atlanta, GA          GA   

   OriginStateFips OriginStateName  DestAirportID  DestAirportSeqID  \
0               12         Florida          12953           1295304   
1               13         Georgia          11697           1169706   
2               12         Florida          10397           1039707   
3               12         Florida          14492           1449202   
4               13         Georgia          12448           1244807   

   DestCityMarketID Dest           DestCityName DestState  DestStateFips  \
0             31703  LGA           New York, NY        NY             36   
1             32467  FLL    Fort Lauderdale, FL        FL             12   
2             30397  ATL            Atlanta, GA        GA             13   
3             34492  RDU     Raleigh/Durham, NC        NC             37   
4             32448  JAN  Jackson/Vicksburg, MS        MS             28   

    DestStateName  CRSDepTime  DepTime  DepDelay  DepDelayMinutes  DepDel15  \
0        New York        1126      NaN       NaN              NaN       NaN   
1         Florida        1631   1627.0      -4.0              0.0       0.0   
2         Georgia        1931   1929.0      -2.0       

Rename syntax for various columns for consistency and accuracy.

In [10]:
# Define the dictionary for renaming columns
rename_dict = {
    'DayofMonth': 'Day_of_Month',
    'DayOfWeek': 'Day_of_Week',
    'FlightDate': 'Flight_Date',
    'Operating_Airline': 'Operating Airline',
    'OriginAirportID': 'Origin_Airport_ID',
    'OriginAirportSeqID': 'Origin_Airport_Seq_ID',
    'OriginCityMarketID': 'Origin_City_Market_ID',
    'OriginCityName': 'Origin_City_Name',
    'OriginState': 'Origin_State',
    'OriginStateFips': 'Origin_State_Fips',
    'OriginStateName': 'Origin_State_Name',
    'DestAirportID': 'Dest_Airport_ID',
    'DestAirportSeqID': 'Dest_Airport_Seq_ID',
    'DestCityMarketID': 'Dest_City_Market_ID',
    'DestCityName': 'Dest_City_Name',
    'DestState': 'Dest_State',
    'DestStateFips': 'Dest_State_Fips',
    'DestStateName': 'Dest_State_Name',
    'CRSDepTime': 'CRS_Dep_Time',
    'DepTime': 'Dep_Time',
    'DepDelay': 'Dep_Delay',
    'DepDelayMinutes': 'Dep_Delay_Minutes',
    'DepDel15': 'Dep_Del_15',
    'DepartureDelayGroups': 'Departure_Delay_Groups',
    'DepTimeBlk': 'Dep_Time_Blk',
    'TaxiOut': 'Taxi_Out',
    'WheelsOff': 'Wheels_Off',
    'WheelsOn': 'Wheels_On',
    'TaxiIn': 'Taxi_In',
    'CRSArrTime': 'CRS_Arr_Time',
    'ArrTime': 'Arr_Time',
    'ArrDelay': 'Arr_Delay',
    'ArrDelayMinutes': 'Arr_Delay_Minutes',
    'ArrDel15': 'Arr_Del_15',
    'ArrivalDelayGroups': 'Arrival_Delay_Groups',
    'ArrTimeBlk': 'Arr_Time_Blk',
    'CancellationCode': 'Cancellation_Code',
    'CRSElapsedTime': 'CRS_Elapsed_Time',
    'ActualElapsedTime': 'Actual_Elapsed_Time',
    'AirTime': 'Air_Time',
    'DistanceGroup': 'Distance_Group',
    'CarrierDelay': 'Carrier_Delay',
    'WeatherDelay': 'Weather_Delay',
    'NASDelay': 'NAS_Delay',
    'SecurityDelay': 'Security_Delay',
    'LateAircraftDelay': 'Late_Aircraft_Delay',
    'FirstDepTime': 'First_Dep_Time',
    'TotalAddGTime': 'Total_Add_G_Time',
    'LongestAddGTime': 'Longest_Add_G_Time',
    'DivDistance': 'Div_Distance'
}

# Rename the columns
df = df.rename(columns=rename_dict)

In [11]:
df.shape

(7013508, 69)

In [12]:
# Drop the additional irrelevant columns - please refer to the 'Cleaning 2019 Flight Data' script for more details.

df = df.drop(columns=['First_Dep_Time', 'Total_Add_G_Time', 'Longest_Add_G_Time', 'Div_Distance'])

Fortunately, this raw data comes with a Duplicate flag column.
I will check the counts for this column to check for any values of 'Y' or Yes.
To be certain, I will also run a duplicate check with the .duplicated command.

In [13]:
df['Duplicate'].value_counts()

Duplicate
N    7013508
Name: count, dtype: int64

In [14]:
# So far so good, now to check for duplicates in the entire dataframe by creating a subset of duplicated rows

# Create a DataFrame containing only the duplicate rows
dups = df[df.duplicated(keep=False)]

# Display the 'dups' DataFrame
print(dups)

Empty DataFrame
Columns: [Year, Quarter, Month, Day_of_Month, Day_of_Week, Flight_Date, Marketing_Airline_Network, DOT_ID_Marketing_Airline, IATA_Code_Marketing_Airline, Flight_Number_Marketing_Airline, IATA_Code_Originally_Scheduled_Code_Share_Airline, Operating_Airline , DOT_ID_Operating_Airline, IATA_Code_Operating_Airline, Tail_Number, Flight_Number_Operating_Airline, Origin_Airport_ID, Origin_Airport_Seq_ID, Origin_City_Market_ID, Origin, Origin_City_Name, Origin_State, Origin_State_Fips, Origin_State_Name, Dest_Airport_ID, Dest_Airport_Seq_ID, Dest_City_Market_ID, Dest, Dest_City_Name, Dest_State, Dest_State_Fips, Dest_State_Name, CRS_Dep_Time, Dep_Time, Dep_Delay, Dep_Delay_Minutes, Dep_Del_15, Departure_Delay_Groups, Dep_Time_Blk, Taxi_Out, Wheels_Off, Wheels_On, Taxi_In, CRS_Arr_Time, Arr_Time, Arr_Delay, Arr_Delay_Minutes, Arr_Del_15, Arrival_Delay_Groups, Arr_Time_Blk, Cancelled, Cancellation_Code, Diverted, CRS_Elapsed_Time, Actual_Elapsed_Time, Air_Time, Flights, Distance,

In [15]:
# No dups!

Now I will address missing and/or null values.

In [16]:
# Set the display option to show all rows
pd.set_option('display.max_rows', None)

In [17]:
# Find the count of null values in each column
null_counts = df.isnull().sum()

In [18]:
# Display the count of null values
print(null_counts)

Year                                                       0
Quarter                                                    0
Month                                                      0
Day_of_Month                                               0
Day_of_Week                                                0
Flight_Date                                                0
Marketing_Airline_Network                                  0
DOT_ID_Marketing_Airline                                   0
IATA_Code_Marketing_Airline                                0
Flight_Number_Marketing_Airline                            0
IATA_Code_Originally_Scheduled_Code_Share_Airline    7012888
Operating_Airline                                          0
DOT_ID_Operating_Airline                                   0
IATA_Code_Operating_Airline                                0
Tail_Number                                            42208
Flight_Number_Operating_Airline                            0
Origin_Airport_ID       

In [19]:
# The IATA_Code_Originally_Scheduled_Code_Share column is nearly completely null
# Given that the Bureau of Transportation Statistics (BTS.gov) stated that this column shouldn't be 
# used for analysis, and it is completely null for 2019, I will be dropping it and keeping the Unique_Code
# column as suggested by BTS.gov
# Here is the language from BTS.gov:
# ""Code assigned by IATA and commonly used to identify a carrier.""
#""As the same code may have been assigned to different carriers over time, the code is not always unique.""
#""For analysis, use the Unique Carrier Code.""

# Drop the 'IATA_Code_Originally_Scheduled_Code_Share_Airline' column
df = df.drop(columns=['IATA_Code_Originally_Scheduled_Code_Share_Airline'])

In [20]:
df.shape

(7013508, 64)

There are a significant amount of null values for Dep_Delay, Taxi_Out, Taxi_In, Arr_Time.
I believe these are null values because the flights were cancelled.
I will check value counts for the Cancelled column. Value counts for '1' should be close if not the same to the values missing.

In [21]:
df['Cancelled'].value_counts()

Cancelled
0.0    6823470
1.0     190038
Name: count, dtype: int64

190,038 flights were cancelled in 2021.

This is fairly close to the missing values in:

Dep_Time                                              185773
Dep_Delay                                             185852
Dep_Delay_Minutes                                     185852
Dep_Del_15                                            185852
Departure_Delay_Groups                                18
Taxi_Out                                              189228
Wheels_Off                                            189228
Wheels_On                                             191747
Taxi_In                                               191
Arr_Time                                              191744
Arr_Delay                                             206702
Arr_Delay_Minutes                                     206702
Arr_Del_15                                            206702
Arrival_Delay_Groups                                  206702747585260014258413

The missing values in:

Cancellation_Code               6823470    6200853    4721342 -- are accounted for because it il flightse 7013508tota6311871ts (50223d flighus t190,038e18ts68234705Theref853 6823470342.6200853ore 4721342 flights have no cancellation code because they were not cancelled.

After careful consideration, I believe it is best to create a cancelled flights flag in order to prevent confusion in future analysis for delays v. cancellation. 

In [22]:
# Identify rows where flights were cancelled

cancelled_flights = df['Cancelled'] == 1

In [23]:
# Create a column to indicate if the data is missing due to cancellation

df['Missing_Due_To_Cancellation'] = cancelled_flights

Now I will address columns containing mixed datatypes.

In [24]:
df.dtypes

Year                                 int64
Quarter                              int64
Month                                int64
Day_of_Month                         int64
Day_of_Week                          int64
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int64
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int64
Operating_Airline                   object
DOT_ID_Operating_Airline             int64
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int64
Origin_Airport_ID                    int64
Origin_Airport_Seq_ID                int64
Origin_City_Market_ID                int64
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                    int64
Origin_State_Name                   object
Dest_Airpor

In [25]:
# I will change many of these datatypes to reduce the amount of memory needed by reducing int64 to int8, 16, or 32 where possible.
# Additionally, I will change the Flight_Date column to a date datatype, instead of an object.
# Lets change Flight_Date first

# Ensure 'Flight_Date' column is in datetime format
df['Flight_Date'] = pd.to_datetime(df['Flight_Date'], errors='coerce')

In [26]:
# Extract only the date part (since we don't have time in the same column)
df['Flight_Date'] = df['Flight_Date'].dt.date

In [27]:
# Verify the result
print(df.head())
print(df.dtypes)

   Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2022        1      1             6            4  2022-01-06   
1  2022        1      1             6            4  2022-01-06   
2  2022        1      1             6            4  2022-01-06   
3  2022        1      1             6            4  2022-01-06   
4  2022        1      1             6            4  2022-01-06   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        DL                     19790   
1                        DL                     19790   
2                        DL                     19790   
3                        DL                     19790   
4                        DL                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             1581   
1                          DL                             1582   
2                          DL                             1582

In [28]:
# Change datatypes to maximize available memory

df = df.astype({
    'Year': 'int16',
    'Quarter': 'int16',
    'Month': 'int16',
    'Day_of_Month': 'int16',
    'Day_of_Week': 'int16',
    'DOT_ID_Marketing_Airline': 'int16',
    'Flight_Number_Marketing_Airline': 'int16',
    'DOT_ID_Operating_Airline': 'int16',
    'Flight_Number_Operating_Airline': 'int16',
    'Origin_Airport_ID': 'int16',
    'Origin_Airport_Seq_ID': 'int16',
    'Origin_City_Market_ID': 'int16',
    'Origin_State_Fips': 'int8',
    'Dest_Airport_ID': 'int16',
    'Dest_Airport_Seq_ID': 'int16',
    'Dest_City_Market_ID': 'int16',
    'Dest_State_Fips': 'int16',
    'CRS_Dep_Time': 'int16'
})

In [29]:
df.dtypes

Year                                 int16
Quarter                              int16
Month                                int16
Day_of_Month                         int16
Day_of_Week                          int16
Flight_Date                         object
Marketing_Airline_Network           object
DOT_ID_Marketing_Airline             int16
IATA_Code_Marketing_Airline         object
Flight_Number_Marketing_Airline      int16
Operating_Airline                   object
DOT_ID_Operating_Airline             int16
IATA_Code_Operating_Airline         object
Tail_Number                         object
Flight_Number_Operating_Airline      int16
Origin_Airport_ID                    int16
Origin_Airport_Seq_ID                int16
Origin_City_Market_ID                int16
Origin                              object
Origin_City_Name                    object
Origin_State                        object
Origin_State_Fips                     int8
Origin_State_Name                   object
Dest_Airpor

In [30]:
df.shape

(7013508, 65)

In [31]:
df.head()

Year  Quarter  Month  Day_of_Month  Day_of_Week Flight_Date  \
0  2022        1      1             6            4  2022-01-06   
1  2022        1      1             6            4  2022-01-06   
2  2022        1      1             6            4  2022-01-06   
3  2022        1      1             6            4  2022-01-06   
4  2022        1      1             6            4  2022-01-06   

  Marketing_Airline_Network  DOT_ID_Marketing_Airline  \
0                        DL                     19790   
1                        DL                     19790   
2                        DL                     19790   
3                        DL                     19790   
4                        DL                     19790   

  IATA_Code_Marketing_Airline  Flight_Number_Marketing_Airline  \
0                          DL                             1581   
1                          DL                             1582   
2                          DL                             1582   
3                          DL                             1583   
4                          DL                             1584   

  Operating_Airline   DOT_ID_Operating_Airline IATA_Code_Operating_Airline  \
0                 DL                     19790                          DL   
1                 DL                     19790                          DL   
2                 DL                     19790                          DL   
3                 DL                     19790                          DL   
4                 DL                     19790                          DL   

  Tail_Number  Flight_Number_Operating_Airline  Origin_Airport_ID  \
0      N315DN                             1581              11697   
1      N545US                             1582              10397   
2      N545US                             1582              11697   
3      N345NB                             1583              11697   
4      N978AT                             1584              10397   

   Origin_Airport_Seq_ID  Origin_City_Market_ID Origin     Origin_City_Name  \
0                  -9942                  32467    FLL  Fort Lauderdale, FL   
1                  -8869                  30397    ATL          Atlanta, GA   
2                  -9942                  32467    FLL  Fort Lauderdale, FL   
3                  -9942                  32467    FLL  Fort Lauderdale, FL   
4                  -8869                  30397    ATL          Atlanta, GA   

  Origin_State  Origin_State_Fips Origin_State_Name  Dest_Airport_ID  \
0           FL                 12           Florida            12953   
1           GA                 13           Georgia            11697   
2           FL                 12           Florida            10397   
3           FL                 12           Florida            14492   
4           GA                 13           Georgia            12448   

   Dest_Airport_Seq_ID  Dest_City_Market_ID Dest         Dest_City_Name  \
0               -15416                31703  LGA           New York, NY   
1                -9942                32467  FLL    Fort Lauderdale, FL   
2                -8869                30397  ATL            Atlanta, GA   
3                 7410               -31044  RDU     Raleigh/Durham, NC   
4                 -377                32448  JAN  Jackson/Vicksburg, MS   

  Dest_State  Dest_State_Fips Dest_State_Name  CRS_Dep_Time  Dep_Time  \
0         NY               36        New York          1126       NaN   
1         FL               12         Florida          1631    1627.0   
2         GA               13         Georgia          1931    1929.0   
3         NC               37  North Carolina          1024    1019.0   
4         MS               28     Mississippi          1117    1113.0   

   Dep_Delay  Dep_Delay_Minutes  Dep_Del_15  Departure_Delay_Groups  \
0        NaN                NaN         NaN                     NaN   
1       -4.0                0.0         0.

In [32]:
# Save Clean 2022 Data

df.to_csv(os.path.join(path, 'clean_flights2022.csv'))